components/eam/src/physics/crm/crm_physics.F90

```diff
@@ -1184,14 +1224,48 @@ subroutine crm_physics_tend(ztodt, state, tend, ptend, pbuf, cam_in, cam_out, &
 
       call crm_history_out(state, ptend, crm_state, crm_rad, crm_output, crm_ecpp_output, qrs, qrl)
 
+      !---------------------------------------------------------------------------------------------
       ! Convert heating rate to Q*dp to conserve energy across timesteps
+      !---------------------------------------------------------------------------------------------
       do m = 1,crm_nz
          k = pver-m+1
          do i = 1,ncol
             crm_rad%qrad(i,:,:,m) = crm_rad%qrad(i,:,:,m) * state%pdel(i,k) ! for energy conservation
          end do
       end do
 
+      !---------------------------------------------------------------------------------------------
+      ! put rad data back in pbuf
+      !---------------------------------------------------------------------------------------------
+      call pbuf_get_field(pbuf, crm_rad_idx_t,    crm_rad_tmp_temperature)
+      call pbuf_get_field(pbuf, crm_rad_idx_qv,   crm_rad_tmp_qv)
+      call pbuf_get_field(pbuf, crm_rad_idx_qc,   crm_rad_tmp_qc)
+      call pbuf_get_field(pbuf, crm_rad_idx_qi,   crm_rad_tmp_qi)
+      call pbuf_get_field(pbuf, crm_rad_idx_cld,  crm_rad_tmp_cld)
+      call pbuf_get_field(pbuf, crm_rad_idx_qrad, crm_rad_tmp_qrad)
+
+      if (MMF_microphysics_scheme .eq. 'm2005') then
+         call pbuf_get_field(pbuf, crm_rad_idx_nc, crm_rad_tmp_nc, start=(/1,1,1,1/), kount=(/pcols,crm_nx_rad, crm_ny_rad, crm_nz/))
+         call pbuf_get_field(pbuf, crm_rad_idx_ni, crm_rad_tmp_ni, start=(/1,1,1,1/), kount=(/pcols,crm_nx_rad, crm_ny_rad, crm_nz/))
+         call pbuf_get_field(pbuf, crm_rad_idx_qs, crm_rad_tmp_qs, start=(/1,1,1,1/), kount=(/pcols,crm_nx_rad, crm_ny_rad, crm_nz/))
+         call pbuf_get_field(pbuf, crm_rad_idx_ns, crm_rad_tmp_ns, start=(/1,1,1,1/), kount=(/pcols,crm_nx_rad, crm_ny_rad, crm_nz/))
+      end if
+
+      do i = 1,ncol
+         crm_rad_tmp_qrad       (i,:,:,:) = crm_rad%qrad       (i,:,:,:)
+         crm_rad_tmp_temperature(i,:,:,:) = crm_rad%temperature(i,:,:,:)
+         crm_rad_tmp_qv         (i,:,:,:) = crm_rad%qv         (i,:,:,:)
+         crm_rad_tmp_qc         (i,:,:,:) = crm_rad%qc         (i,:,:,:)
+         crm_rad_tmp_qi         (i,:,:,:) = crm_rad%qi         (i,:,:,:)
+         crm_rad_tmp_cld        (i,:,:,:) = crm_rad%cld        (i,:,:,:)
+         if (MMF_microphysics_scheme .eq. 'm2005') then
```

Comment:
I think it's probably better to move the logical out of the loop here and just have two loops. Maybe combine with the other logical block above that gets the fields?

Reply:
Why would that be better?

Reply:
Performance, but probably negligible here.